### PR TITLE
Allow broadcasting of value x params in Categorical

### PR DESCRIPTION
**Pull request by neerajprad**
Allows for broadcasting of values and parameters across the batch dimensions for Categorical. The other distributions do not need to do explicit broadcasting in `log_prob` and already handle this case.

Fixes #79.

### CODE REVIEW

**Comment by neerajprad** [comment]
cc. @fritzo, @alicanb, @apaszke

**Review by fritzo — APPROVED** [review]
Thanks for fixing this.

**Comment by neerajprad** [comment]
Actually, given that we need to make changes in #4624, I am reverting the last commit, so that we can merge this and resolve all the probs/logits lazy evaluation in one pass in a separate PR.

**Comment by neerajprad** [comment]
This should be good to merge as is. Will resolve all issues with lazy evaluation of probs/logits for all distributions in a separate PR.

**Review by apaszke — APPROVED** [review]
